src/main/scala/io/iohk/ethereum/vm/ProgramContext.scala

```diff
@@ -6,40 +6,33 @@ import io.iohk.ethereum.domain._
 object ProgramContext {
   def apply[W <: WorldStateProxy[W, S], S <: Storage[S]](
     stx: SignedTransaction,
+    recipientAddress: Address,
     blockHeader: BlockHeader,
     world: W,
     config: EvmConfig): ProgramContext[W, S] = {
 
     import stx.tx
 
     val senderAddress = stx.senderAddress
-    val (world1, recipientAddress, program) = callOrCreate[W, S](world, tx, senderAddress)
+    val program = getProgram[W, S](world, tx)
```

Comment:
There's still logic here, deciding whether it's message call or contract creation, calculating gas, obtaining the code...
Ideally all this logic should be moved to the ledger.